README.md
# DEPRECATED

Can be replaced with: [i18next-locize-backend](https://github.com/locize/i18next-locize-backend)

---

# Introduction

This is a i18next backend to be used with node.js for the [locize](http://locize.com) service. It's for the node.js server what the [i18next-locize-backend](https://github.com/locize/i18next-locize-backend) is for the browser.

# Getting started

Source can be loaded via [npm](https://www.npmjs.com/package/i18next-node-locize-backend).

```
$ npm install i18next-node-locize-backend
```

Wiring up:

```js
var i18next = require('i18next');
var Backend = require('i18next-node-locize-backend');

i18next
  .use(Backend)
  .init(i18nextOptions);
```

As with all modules you can either pass the constructor function (class) to the i18next.use or a concrete instance.

## Backend Options

```js
{
  // the id of your locize project
  projectId: '[PROJECTID]',

  // add an api key if you want to send missing keys
  apiKey: '[APIKEY]',

  // the reference language of your project
  referenceLng: '[LNG]',

  // version - defaults to latest
  version: '[VERSION]',

  // private - set to true if you version on locize is set to use private publish
  private: false
}
```

Options can be passed in:

**preferred** - by setting options.backend in i18next.init:

```js
var i18next = require('i18next');
var Backend = require('i18next-node-locize-backend');

i18next
  .use(Backend)
  .init({
    backend: options
  });
```

on construction:

```js
var Backend = require('i18next-node-locize-backend');
var backend = new Backend(options);
```

by calling init:

```js
var Backend = require('i18next-node-locize-backend');
var backend = new Backend();
backend.init(options);
```


## Additional API endpoints

### backend.getLanguages

Will return a list of all languages in your project including percentage of translations done per version.

```js
import Backend from 'i18next-node-locize-backend';
const backend = new Backend(options);

backend.getLanguages((err, data) => {
  /*
  data is:

  {
    "en": {
      "name": "English",
      "nativeName": "English",
      "isReferenceLanguage": true,
      "translated": {
        "latest": 1
      }
    },
    "de": {
      "name": "German",
      "nativeName": "Deutsch",
      "isReferenceLanguage": false,
      "translated": {
        "latest": 0.9
      }
    }
  }
  */
});

// or
i18next.services.backendConnector.backend.getLanguages(callback);
```

### backend.getOptions

Will return an object containing useful informations for the i18next init options.

```js
import Backend from 'i18next-node-locize-backend';
const backend = new Backend(options);

backend.getOptions((err, data) => {
  /*
  data is:

  {
    fallbackLng: 'en',
    referenceLng: 'en',
    whitelist: ['en', 'de'],
    load: 'languageOnly|all' // depending on your whitelist has locals having region like en-US
  }
  */
});

// or
i18next.services.backendConnector.backend.getOptions(callback);
```

You can set a threshold for languages to be added to whitelist by setting whitelistThreshold in backend options (eg: 1 = 100% translated, 0.9 = 90% translated).

## SPECIAL - let the backend determine some options to improve loading

You can load some information from the backend to eg. set whitelist for i18next just supporting languages you got in your locize project.

You will get i18next options for (same as above backend.getOptions):

- fallbackLng
- whitelist
- load

```js
import i18next from 'i18next';
import Backend from 'i18next-node-locize-backend';

const backend = new Backend({
  projectId: '[PROJECTID]',
  apiKey: '[APIKEY]',
  version: '[VERSION]',
  // referenceLng -> not needed as will be loaded from API
}, (err, opts) => {
  i18next
    .use(backend)
    .init({ ...opts, ...yourOptions}); // yourOptions should not include backendOptions!
});
```

## IMPORTANT ADVICE FOR SERVERLESS environments - AWS lambda, Google Cloud Functions, Azure Functions, etc...

<font color="red">
  <b>Please be aware</b>
</font>

Due to how serverless functions work, you cannot guarantee that a cached version of your data is available. Serverless functions are short-lived, and can shut down at any time, purging any in-memory or filesystem cache. This may be an acceptable trade-off, but sometimes it isn't acceptable.

**Because of this we suggest to download the translations in your CI/CD pipeline (via [cli](https://github.com/locize/locize-cli#download-current-published-files) or via [api](https://docs.locize.com/integration/api#list-all-namespace-resources)) and package them with your serverless function.**

### For example with [i18next-node-fs-backend](https://github.com/i18next/i18next-node-fs-backend)

```js
import i18next from 'i18next';
import Backend from 'i18next-node-fs-backend';

const backend = new Backend({
  // path where resources get loaded from
  loadPath: '/locales/{{lng}}/{{ns}}.json'
});

i18next
  .use(backend)
  .init({ ...opts, ...yourOptions}); // yourOptions should not include backendOptions!
```

#### Another example with [i18next-sync-fs-backend](https://github.com/sallar/i18next-sync-fs-backend): [here](https://stackoverflow.com/questions/59591125/initialize-i18next-in-azure-function/59634847#59634847)

### or just [import/require](https://www.i18next.com/how-to/add-or-load-translations#add-on-init) your files directly

```js
import i18next from 'i18next';
import en from './locales/en.json'
import de from './locales/de.json'

i18next
  .init({
    ...opts,
    ...yourOptions,
    resources: {
      en,
      de
    }
  });
```


CHANGELOG.md
### 3.1.2

- page missing request

### 3.1.1

- make sure checkIfProjectExists is not called, while waiting for response

### 3.1.0

- check if project exists

### 3.0.1

- retry on 400, 408 status

### 3.0.0

- using the new `locize.app` domain replacing the deprecated `locize.io` domain

### 2.4.4

- warn for missing options

### v2.4.3

- update deps [6](https://github.com/locize/i18next-node-locize-backend/pull/6)

### v2.4.2

- clear existing interval on init call if existing - make new one (or not if set false)

### v2.4.1

- fixes issue with payloads

### v2.4.0

- support for private versions

### v2.3.0

- support for update route

### v2.2.2

- fixes backendConnector call (setting services on init)

### v2.2.1

- fix getLanguages route to return error on not getting data

### v2.2.0

- adds getOptions route
- add special construction with callback to autoload possible i18next options.

### v2.1.1

- fixes debounced write to send every namespace-language pair

### v2.1.0

- supports submitting options tDescription (context) i18next >= 10.4.1

### v2.0.0

- update dependencies
- update build process - remove usage of gulp


src/utils.js
export function debounce(func, wait, immediate) {
	var timeout;
	return function() {
		var context = this, args = arguments;
		var later = function() {
			timeout = null;
			if (!immediate) func.apply(context, args);
		};
		var callNow = immediate && !timeout;
		clearTimeout(timeout);
		timeout = setTimeout(later, wait);
		if (callNow) func.apply(context, args);
	};
};

function getLastOfPath(object, path, Empty) {
  function cleanKey(key) {
    return (key && key.indexOf('###') > -1) ? key.replace(/###/g, '.') : key;
  }

  let stack = (typeof path !== 'string') ? [].concat(path) : path.split('.');
  while(stack.length > 1) {
    if (!object) return {};

    let key = cleanKey(stack.shift());
    if (!object[key] && Empty) object[key] = new Empty();
    object = object[key];
  }

  if (!object) return {};
  return {
    obj: object,
    k: cleanKey(stack.shift())
  };
}

export function setPath(object, path, newValue) {
  let { obj, k } = getLastOfPath(object, path, Object);

  obj[k] = newValue;
}

export function pushPath(object, path, newValue, concat) {
  let { obj, k } = getLastOfPath(object, path, Object);

  obj[k] = obj[k] || [];
  if (concat) obj[k] = obj[k].concat(newValue);
  if (!concat) obj[k].push(newValue);
}

export function getPath(object, path) {
  let { obj, k } = getLastOfPath(object, path);

  if (!obj) return undefined;
  return obj[k];
}



const regexp = new RegExp('\{\{(.+?)\}\}', 'g');

function makeString(object) {
  if (object == null) return '';
  return '' + object;
}

export function interpolate(str, data, lng) {
  let match, value;

  function regexSafe(val) {
    return val.replace(/\$/g, '$$$$');
  }

  // regular escape on demand
  while(match = regexp.exec(str)) {
     value = match[1].trim();
     if (typeof value !== 'string') value = makeString(value);
     if (!value) value = '';
     value = regexSafe(value);
     str = str.replace(match[0], data[value] || value);
     regexp.lastIndex = 0;
  }
  return str;
}


export function isMissingOption(obj, props) {
  return props.reduce((mem, p) => {
    if (mem) return mem;
    if (!obj || !obj[p] || typeof obj[p] !== 'string' || !obj[p].toLowerCase() === p.toLowerCase()) {
      const err = `i18next-node-locize-backend :: got "${obj[p]}" in options for ${p} which is invalid.`;
      console.warn(err);
      return err;
    }
    return false;
  }, false)
}

export function optionExist(obj, props) {
  return !isMissingOption(obj, props);
}

src/index.js
import * as utils from './utils';
import request from 'request';

function ajax(url, callback, data) {
  if (data) {
    let reqOptions = typeof url === 'string' ? { uri: url, body: data, json: true } : { ...url, ...{ body: data, json: true }};
    request.post(reqOptions, function(err, res, body) {
      if (err) console.log(err);
      callback(err, body, res);
    });
  } else {
    request(url, function(err, res, body) {
      if (err) console.log(err);
      callback(err, body, res);
    });
  }
};

function getDefaults() {
  return {
    loadPath: 'https://api.locize.app/{{projectId}}/{{version}}/{{lng}}/{{ns}}',
    privatePath: 'https://api.locize.app/private/{{projectId}}/{{version}}/{{lng}}/{{ns}}',
    getLanguagesPath: 'https://api.locize.app/languages/{{projectId}}',
    addPath: 'https://api.locize.app/missing/{{projectId}}/{{version}}/{{lng}}/{{ns}}',
    updatePath: 'https://api.locize.app/update/{{projectId}}/{{version}}/{{lng}}/{{ns}}',
    referenceLng: 'en',
    version: 'latest',
    private: false,
    whitelistThreshold: 0.9,
    reloadInterval: 60 * 60 * 1000,
    checkForProjectTimeout: 3 * 1000
  };
}

class Backend {
  constructor(services, options = {}, allOptions = {}, callback) {
    if (services && services.projectId) {
      this.init(null, services, allOptions, options);
    } else {
      this.init(services, options, allOptions, callback);
    }

    this.type = 'backend';
  }

  init(services, options = {}, allOptions = {}, callback) {
    this.services = services;
    this.options = {...getDefaults(), ...this.options, ...options}; // initial
    this.allOptions = allOptions;
    this.somethingLoaded = false;
    this.isProjectNotExisting = false;

    if (typeof callback === 'function') {
      this.getOptions((err, opts) => {
        if (err) return callback(err);

        this.options.referenceLng = options.referenceLng || opts.referenceLng || this.options.referenceLng;
        callback(null, opts);
      });
    }

    this.queuedWrites = { pending: {} };
    this.debouncedProcess = utils.debounce(this.process, 10000);

    if (this.interval) {
      clearInterval(this.interval);
    }
    if (this.options.reloadInterval) {
      this.interval = setInterval(() => {
        this.reload();
      }, this.options.reloadInterval);
    }
  }

  reload() {
    const { backendConnector, resourceStore, languageUtils, logger } = this.services;
    if (!backendConnector) return;

    const currentLanguage = backendConnector.language;
    if (currentLanguage && currentLanguage.toLowerCase() === 'cimode') return; // avoid loading resources for cimode

    let toLoad = [];

    let append = lng => {
      let lngs = languageUtils.toResolveHierarchy(lng);
      lngs.forEach(l => {
        if (toLoad.indexOf(l) < 0) toLoad.push(l);
      });
    };

    append(currentLanguage);

    if (this.allOptions.preload) {
      this.allOptions.preload.forEach(l => {
        append(l);
      });
    }

    toLoad.forEach(lng => {
      this.allOptions.ns.forEach(ns => {
        backendConnector.read(lng, ns, 'read', null, null, (err, data) => {
          if (err) logger.warn(`loading namespace ${ns} for language ${lng} failed`, err);
          if (!err && data) logger.log(`loaded namespace ${ns} for language ${lng}`, data);

          backendConnector.loaded(`${lng}|${ns}`, err, data);
        });
      });
    });
  }

  getLanguages(callback) {
    const isMissing = utils.isMissingOption(this.options, ['projectId']);
    if (isMissing) return callback(new Error(isMissing));

    let url = utils.interpolate(this.options.getLanguagesPath, { projectId: this.options.projectId });

    if (this.isProjectNotExisting) return callback(new Error(`locize project ${this.options.projectId} does not exist!`));

    this.loadUrl(url, (err, ret, info) => {
      if (!this.somethingLoaded && info && info.resourceNotExisting) {
        this.isProjectNotExisting = true;
        return callback(new Error(`locize project ${this.options.projectId} does not exist!`));
      }
      this.somethingLoaded = true;
      callback(err, ret)
    });
  }

  getOptions(callback) {
    this.getLanguages((err, data) => {
      if (err) return callback(err);

      const keys = Object.keys(data);
      if (!keys.length) return callback(new Error('was unable to load languages via API'));

      const referenceLng = keys.reduce((mem, k) => {
        const item = data[k];
        if (item.isReferenceLanguage) mem = k;
        return mem;
      }, '');

      const whitelist = keys.reduce((mem, k) => {
        const item = data[k];
        if (item.translated[this.options.version] && item.translated[this.options.version] >= this.options.whitelistThreshold) mem.push(k)
        return mem;
      }, []);

      const hasRegion = keys.reduce((mem, k) => {
        if (k.indexOf('-') > -1) return true;
        return mem;
      }, false);

      callback(null, {
        fallbackLng: referenceLng,
        referenceLng,
        whitelist,
        load: hasRegion ? 'all' : 'languageOnly'
      });
    });
  }

  checkIfProjectExists(callback) {
    const { logger } = this.services;
    if (this.somethingLoaded) {
      if (callback) callback(null);
      return;
    }
    if (this.alreadyRequestedCheckIfProjectExists) {
      setTimeout(() => this.checkIfProjectExists(callback), this.options.checkForProjectTimeout);
      return;
    }
    this.alreadyRequestedCheckIfProjectExists = true;
    this.getLanguages((err) => {
      if (err && err.message && err.message.indexOf('does not exist') > 0) {
        if (callback) return callback(err);
        logger.error(err.message);
      }
    });
  }

  read(language, namespace, callback) {
    const { logger } = this.services || { logger: console };
    let url;
    if (this.options.private) {
      const isMissing = utils.isMissingOption(this.options, ['projectId', 'version', 'apiKey'])
      if (isMissing) return callback(new Error(isMissing), false);

      url = {
        uri: utils.interpolate(this.options.privatePath, { lng: language, ns: namespace, projectId: this.options.projectId, version: this.options.version }),
        headers: {
          'Authorization': this.options.apiKey
        }
      };
    } else {
      const isMissing = utils.isMissingOption(this.options, ['projectId', 'version'])
      if (isMissing) return callback(new Error(isMissing), false);

      url = utils.interpolate(this.options.loadPath, { lng: language, ns: namespace, projectId: this.options.projectId, version: this.options.version });
    }

    if (this.isProjectNotExisting) {
      const err = new Error(`locize project ${this.options.projectId} does not exist!`);
      logger.error(err.message);
      if (callback) callback(err);
      return;
    }

    this.loadUrl(url, (err, ret, info) => {
      if (!this.somethingLoaded) {
        if (info && info.resourceNotExisting) {
          setTimeout(() => this.checkIfProjectExists(), this.options.checkForProjectTimeout);
        } else {
          this.somethingLoaded = true;
        }
      }
      callback(err, ret)
    });
  }

  loadUrl(url, callback) {
    ajax(url, (err, data, res) => {
      if (err) return callback(err, true); // retry

      const statusCode = res.statusCode;
      const resourceNotExisting = res.headers['x-cache'] === 'Error from cloudfront'
      if (statusCode && (statusCode === 408 || statusCode === 400)) return callback('failed loading ' + url, true /* retry */, { resourceNotExisting });
      if (statusCode && statusCode >= 500 && statusCode < 600) return callback('failed loading ' + url, true /* retry */, { resourceNotExisting });
      if (statusCode && statusCode >= 400 && statusCode < 500) return callback('failed loading ' + url, false /* no retry */, { resourceNotExisting });

      let ret;
      try {
        ret = JSON.parse(data);
      } catch (e) {
        err = 'failed parsing ' + url + ' to json';
      }
      if (err) return callback(err, false, { resourceNotExisting });
      callback(null, ret, { resourceNotExisting });
    });
  }

  create(languages, namespace, key, fallbackValue, callback, options) {
    this.checkIfProjectExists((err) => {
      if (err) {
        if (callback) callback(err);
        return;
      }

      // missing options
      const isMissing = utils.isMissingOption(this.options, ['projectId', 'version', 'apiKey', 'referenceLng'])
      if (isMissing) return callback(new Error(isMissing));

      if (typeof languages === 'string') languages = [languages];

      languages.forEach(lng => {
        if (lng === this.options.referenceLng) this.queue.call(this, this.options.referenceLng, namespace, key, fallbackValue, callback, options);
      });
    });
  }

  writePage(lng, namespace, missings, callback) {
    let missingUrl = utils.interpolate(this.options.addPath, { lng: lng, ns: namespace, projectId: this.options.projectId, version: this.options.version });
    let updatesUrl = utils.interpolate(this.options.updatePath, { lng: lng, ns: namespace, projectId: this.options.projectId, version: this.options.version });

    let hasMissing= false;
    let hasUpdates = false;
    const payloadMissing = {};
    const payloadUpdate = {};

    missings.forEach(item => {
      const value = (item.options && item.options.tDescription) ? { value: item.fallbackValue || '', context: { text: item.options.tDescription } } : item.fallbackValue || ''
      if (item.options && item.options.isUpdate) {
        if (!hasUpdates) hasUpdates = true;
        payloadUpdate[item.key] = value;
      } else {
        if (!hasMissing) hasMissing = true;
        payloadMissing[item.key] = value;
      }
    });

    let todo = 0;
    if (hasMissing) todo++;
    if (hasUpdates) todo++;
    const doneOne = () => {
      todo--;
      if (!todo) callback()
    };

    if (!todo) doneOne();

    if (hasMissing) {
      const reqOptions = {
        uri: missingUrl,
        headers: {
          'Authorization': this.options.apiKey
        }
      };
      ajax(reqOptions, (err, payloadMissing, res) => {
        //const statusCode = xhr.status.toString();
        // TODO: if statusCode === 4xx do log

        doneOne();
      }, payloadMissing);
    }

    if (hasUpdates) {
      const reqOptions = {
        uri: updatesUrl,
        headers: {
          'Authorization': this.options.apiKey
        }
      };
      ajax(reqOptions, (err, payloadUpdate, res) => {
        //const statusCode = xhr.status.toString();
        // TODO: if statusCode === 4xx do log

        doneOne();
      }, payloadUpdate);
    }
  }

  write(lng, namespace) {
    let lock = utils.getPath(this.queuedWrites, ['locks', lng, namespace]);
    if (lock) return;

    let missings = utils.getPath(this.queuedWrites, [lng, namespace]);
    utils.setPath(this.queuedWrites, [lng, namespace], []);
    const pageSize = 1000;

    if (missings.length) {
      // lock
      utils.setPath(this.queuedWrites, ['locks', lng, namespace], true);

      const namespaceSaved = () => {
        // unlock
        utils.setPath(this.queuedWrites, ['locks', lng, namespace], false);

        missings.forEach((missing) => {
          if (missing.callback) missing.callback();
        });

        // rerun
        this.debouncedProcess(lng, namespace);
      };

      const amountOfPages = missings.length / pageSize;
      let pagesDone = 0;

      let page = missings.splice(0, pageSize);
      this.writePage(lng, namespace, page, () => {
        pagesDone++;
        if (pagesDone >= amountOfPages) namespaceSaved();
      });
      while (page.length === pageSize) {
        page = missings.splice(0, pageSize);
        if (page.length) this.writePage(lng, namespace, page, () => {
          pagesDone++;
          if (pagesDone >= amountOfPages) namespaceSaved();
        });
      }
    }
  }

  process() {
    Object.keys(this.queuedWrites).forEach((lng) => {
      if (lng === 'locks') return;
      Object.keys(this.queuedWrites[lng]).forEach((ns) => {
        const todo = this.queuedWrites[lng][ns];
        if (todo.length) {
          this.write(lng, ns);
        }
      });
    });
  }

  queue(lng, namespace, key, fallbackValue, callback, options) {
    utils.pushPath(this.queuedWrites, [lng, namespace], {key: key, fallbackValue: fallbackValue || '', callback: callback, options});

    this.debouncedProcess();
  }
}

Backend.type = 'backend';


export default Backend;
